Make postprocessors work when using complex-valued solutions.

bangerth opened the pull request:
It turned out that I had this patch sitting in my repository for more than a year now. It is necessary for the step-58 program that I have yet to finish.

Passes all of my tests. I have a feeling that someone is going to ask me for a test but wanted to put this out there for the discussion in #8625. Would appreciate feedback.

bangerth commented:
I've re-thought this patch. I think my original approach was wrong, and what is really the case is that we have never actually tested complex-valued solutions with `DataOut` postprocessors at all. So I've rewritten the patch so that it simply rules that out with an assertion. This passes all tests (because we don't have any tests that complex-valued solutions, `DataOut`, and postprocessors at the same time), and makes sure that we exit with an error early on, rather than doing something silly downstream.

bangerth commented:
I'll try to fix the actual omission of the functionality in a follow-up patch. 
(FYI: @dangars)

/rebuild

dangars commented on source/numerics/data_out.cc:
```diff
@@ -200,35 +210,231 @@ DataOut<dim, DoFHandlerType>::build_one_patch(
-                      scratch_data.patch_values_system.solution_hessians);
+                  scratch_data.resize_system_vectors(
+                    n_components *
+                    (this->dof_data[dataset]->is_complex_valued() ? 2 : 1));
```
@bangerth If I understand correctly. If the solution is a scalar value, I will get 2 components in DataPostprocessor (real and imag). If the solution is a complex vector with 3 components, I will get 6 components in DataPostprocessor that contain the real and the imaginary parts. Is this correct?

That is why you mentioned in #8625 that is possible to calculate complex derive quantities because you have access to all the real and imaginary components at the same time in DataPostprocessor. I realized that I misunderstood :)

bangerth commented:
On 9/5/19 3:15 AM, Daniel Garcia-Sanchez wrote:
> @bangerth <https://github.com/bangerth> If I understand correctly. If the 
> solution is a scalar value, I will get 2 components in DataPostprocessor (real 
> and imag). If the solution is a complex vector with 3 components, I will get 6 
> components in DataPostprocessor that contain the real and the imaginary parts. 
> Is this correct?

Yes, correct. Except that I haven't implemented all of this yet, but that's 
the plan at least.

dangars commented:
> Yes, correct. Except that I haven't implemented all of this yet, but that's the plan at least.

This sounds great! thanks :)

bangerth commented:
I've updated this patch and it works now for postprocessors and complex-valued solutions -- for the moment only for scalar solutions, but that will be fixed in follow-ups.

The key to understanding most of the patch is to realize the following two issues:
* When using a postprocessor, *all* results are real-valued, regardless of whether the inputs are real- or complex-valued. To fix the fact that I had not originally accounted for this required the addition of a fair number of `&& (this->dof_data[dataset]->postprocessor == nullptr)` conditions when we ask whether the input is complex-valued.
* We need to have code that extracts both the real and imaginary parts of a complex-valued solution, and puts these things into the `DataPostprocessorInputs` structure. This is repetitive, but I don't see a better solution to the problem -- we have to somehow pick the various parts apart. The way I currently do this would clearly be very onerous if anyone ever thought about using octonions as the elements of some input vector, but I feel that this is not likely to happen anytime soon.

I've added a test that verifies that everything we do here works as intended -- which, indeed, it does!

bangerth commented:
Ping?

bangerth commented:
Ping?

bangerth commented:
@kronbichler -- what, not a single comment? Have I started writing clean code finally? ;-)

kronbichler commented:
I am maybe not the most critical reviewer, but the code seemed very clean to me. :wink:

bangerth commented:
I take my compliments where I can get them ;-)
